chain/namada/namada_node.go

```diff
@@ -0,0 +1,334 @@
+
+	nv, err := dockerClient.VolumeCreate(ctx, volumetypes.CreateOptions{
+		Labels: map[string]string{
+			dockerutil.CleanupLabel:   testName,
+			dockerutil.NodeOwnerLabel: nn.Name(),
+		},
+	})
+	if err != nil {
+		return nil, fmt.Errorf("creating namada volume: %w", err)
+	}
+
+	nn.VolumeName = nv.Name
+	if err := dockerutil.SetVolumeOwner(ctx, dockerutil.VolumeOwnerOptions{
+		Log: log,
+
+		Client: dockerClient,
+
+		VolumeName: nn.VolumeName,
+		ImageRef:   nn.Image.Ref(),
+		TestName:   nn.TestName,
+		UidGid:     image.UIDGID,
+	}); err != nil {
+		return nil, fmt.Errorf("set namada volume owner: %w", err)
+	}
+
+	return nn, nil
+}
+
+func (n *NamadaNode) Exec(ctx context.Context, cmd []string, env []string) ([]byte, []byte, error) {
+	job := dockerutil.NewImage(n.logger(), n.DockerClient, n.NetworkID, n.TestName, n.Image.Repository, n.Image.Version)
+	opts := dockerutil.ContainerOptions{
+		Env:   env,
+		Binds: n.Bind(),
+	}
+	res := job.Run(ctx, cmd, opts)
+	return res.Stdout, res.Stderr, res.Err
+}
+
+func (n *NamadaNode) logger() *zap.Logger {
+	return n.log.With(
+		zap.String("chain_id", n.Chain.Config().ChainID),
+		zap.String("test", n.TestName),
+	)
+}
+
+// Name of the test node container.
+func (n *NamadaNode) Name() string {
+	return fmt.Sprintf("%s-%s-%d-%s", n.Chain.Config().ChainID, n.NodeType(), n.Index, n.TestName)
+}
+
+func (n *NamadaNode) HostName() string {
+	return dockerutil.CondenseHostName(n.Name())
+}
+
+func (n *NamadaNode) Bind() []string {
+	return []string{fmt.Sprintf("%s:%s", n.VolumeName, n.HomeDir())}
+}
+
+// Home directory in the Docker filesystem.
+func (n *NamadaNode) HomeDir() string {
+	return "/home/namada"
+}
+
+func (n *NamadaNode) NodeType() string {
+	nodeType := "fn"
+	if n.Validator {
+		nodeType = "val"
+	}
+	return nodeType
+}
+
+func (n *NamadaNode) NewRPCClient(addr string) error {
+	httpClient, err := libclient.DefaultHTTPClient(addr)
+	if err != nil {
+		return err
+	}
+
+	httpClient.Timeout = 10 * time.Second
+	rpcClient, err := rpchttp.NewWithClient(addr, "/websocket", httpClient)
+	if err != nil {
+		return err
+	}
+
+	n.Client = rpcClient
+	return nil
+}
+
+func (n *NamadaNode) Height(ctx context.Context) (int64, error) {
+	stat, err := n.Client.Status(ctx)
+	if err != nil {
+		return 0, fmt.Errorf("tendermint client status: %w", err)
+	}
+	return stat.SyncInfo.LatestBlockHeight, nil
+}
+
+func (n *NamadaNode) CreateContainer(ctx context.Context) error {
+	setConfigDir := fmt.Sprintf("NAMADA_NETWORK_CONFIGS_DIR=%s", n.HomeDir())
+
+	joinNetworkCmd := fmt.Sprintf(`%s namadac --base-dir %s utils join-network --add-persistent-peers --chain-id %s --allow-duplicate-ip`, setConfigDir, n.HomeDir(), n.Chain.Config().ChainID)
+	if n.Validator {
+		joinNetworkCmd += " --genesis-validator " + fmt.Sprintf("validator-%d", n.Index)
+	}
+
+	updateCmd := fmt.Sprintf(`sed -i s/127.0.0.1:26657/0.0.0.0:26657/g %s/%s/config.toml`, n.HomeDir(), n.Chain.Config().ChainID)
```

Comment:
There is a `testutil.ModifyTomlConfigFile` in `"github.com/strangelove-ventures/interchaintest/v8/testutil"` that assist with updating toml files if desired

Reply:
Thanks! I replaced it.